File: lib/steps/s05b_gender_detect.py
import json
from pathlib import Path
from typing import Any, Dict

from core.step_base import StepBase
from utils.gender_detector import GenderDetector


class StepGenderDetect(StepBase):
    step_id = "s05b_gender_detect"
    depends_on = ["s04_audio_separate", "s05_asr"]

    def run(self, workspace: Path, config: Dict[str, Any], job_state: Any) -> Dict[str, Any]:
        asr_info = job_state.get_step_output("s05_asr") or {}
        asr_file = Path(asr_info.get("transcript_file", workspace / "s05_asr.json"))

        segments = []
        if asr_file.exists():
            with open(asr_file, "r", encoding="utf-8") as f:
                segments = json.load(f)

        gender_map = {}
        enable_gender = config.get("enable_gender_tts", False)

        if enable_gender:
            audio_info = job_state.get_step_output("s04_audio_separate") or {}
            voice_path = Path(audio_info.get("voice", workspace / "vocals.wav"))

            for idx, seg in enumerate(segments):
                seg_id = str(seg.get("id", idx))
                start = float(seg.get("start", 0.0))
                end = float(seg.get("end", start + 1.0))

                gender = GenderDetector.estimate_segment_f0(voice_path, start, end)
                gender_map[seg_id] = {
                    "start": start,
                    "end": end,
                    "text": seg.get("text", ""),
                    "gender": gender
                }
        else:
            for idx, seg in enumerate(segments):
                seg_id = str(seg.get("id", idx))
                gender_map[seg_id] = {
                    "start": float(seg.get("start", 0.0)),
                    "end": float(seg.get("end", 0.0)),
                    "gender": "unknown"
                }

        out_file = workspace / "s05b_gender.json"
        with open(out_file, "w", encoding="utf-8") as f:
            json.dump(gender_map, f, ensure_ascii=False, indent=2)

        return {
            "gender_file": str(out_file),
            "enabled": enable_gender,
            "detected_count": len(gender_map)
        }

File: video_engine/lib/steps/s05b_gender_detect.dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as p;
import '../core/job_state.dart';
import '../core/step_base.dart';
import '../ffi/audio_dsp_bindings.dart';

class StepGenderDetect extends StepBase {
  @override
  String get stepId => 's05b_gender_detect';

  @override
  List<String> get dependsOn => const ['s04_audio_separate', 's05_asr'];

  @override
  List<String> get stepConfigKeys => const ['enable_gender_tts', 'ocr_only'];

  @override
  Future<Map<String, dynamic>> run(
    Directory workspace,
    Map<String, dynamic> config,
    JobState jobState,
  ) async {
    final outFile = File(p.join(workspace.path, 's05b_gender.json'));

    final isOcrOnly = config['ocr_only'] == true;
    final enableGender = config['enable_gender_tts'] == true;

    if (isOcrOnly || !enableGender) {
      outFile.writeAsStringSync('{}');
      return {
        'skipped': true,
        'gender_file': outFile.path,
      };
    }

    final asrInfo = jobState.getStepOutput('s05_asr') ?? {};
    final audioInfo = jobState.getStepOutput('s04_audio_separate') ?? {};

    final voicePath = audioInfo['voice'] as String?;
    final transcriptFile = asrInfo['transcript_file'] as String?;

    if (voicePath == null || transcriptFile == null || !File(voicePath).existsSync() || !File(transcriptFile).existsSync()) {
      outFile.writeAsStringSync('{}');
      return {'skipped': true, 'gender_file': outFile.path};
    }

    final segments = (jsonDecode(File(transcriptFile).readAsStringSync()) as List)
        .map((e) => Map<String, dynamic>.from(e as Map))
        .toList();

    final genderMap = <String, Map<String, dynamic>>{};

    for (int i = 0; i < segments.length; i++) {
      final seg = segments[i];
      final start = double.tryParse(seg['start']?.toString() ?? '0.0') ?? 0.0;
      final end = double.tryParse(seg['end']?.toString() ?? '${start + 1.0}') ?? (start + 1.0);
      final segId = seg['id']?.toString() ?? i.toString();

      final gender = AudioDspBindings.estimateGender(
        audioPath: voicePath,
        startSec: start,
        endSec: end,
      );

      genderMap[segId] = {
        'id': segId,
        'start': start,
        'end': end,
        'gender': gender,
      };
    }

    outFile.writeAsStringSync(
      const JsonEncoder.withIndent('  ').convert(genderMap),
      flush: true,
    );

    return {
      'gender_file': outFile.path,
      'detected_count': genderMap.length,
    };
  }
}
